### PR TITLE
Fix lookup precedence

### PR DESCRIPTION
**Pull request by ulyssessouza**

### CODE REVIEW

**Review by nicksieger — APPROVED** [review]
LGTM

File: dotenv/godotenv_test.go
Comment (milas) [inline dotenv/godotenv_test.go:626]: Same as above - this seems incorrect now. The test is actually called `TestSubstitutionsWithShellEnvPrecedence`, but now we're asserting that env file values have higher precedence 🤔

File: cli/options.go
Comment (ulyssessouza) [reply cli/options.go:256]: No. That's exactly the intention.
If the users want to choose the OS one they can use something like this in the `.env` file:
```
MYVAR=${MYVAR:-DefaultValue}
```
This would use the OS variable when present and use `DefaultValue` when unset or empty.

That's why I had to fix the [default values](https://github.com/compose-spec/compose-go/pull/276) first.

File: dotenv/godotenv_test.go
Comment (ulyssessouza) [reply dotenv/godotenv_test.go:626]: I also added more tests with default values to show the trick.

**Review by glours — APPROVED** [review]
LGTM
Waiting for @milas approval before merging

File: cli/options.go
Comment (milas) [reply cli/options.go:256]: Ahhh okay I just did some sleuthing, and v1 _does_ behave in the same way that you are making the changes here: env has higher precedence than shell, so interpolation/default syntax is required for any values you want to "fall through" to shell.

HOWEVER, my confusion came from the fact that the Compose docs, even historically going back to v1, actually state the opposite behavior 🙈  (e.g. here's the page in 2019 -> https://github.com/docker/docker.github.io/blob/f2093d4e2a5f950d1fd8e6dc61441a1d90cecdb4/compose/environment-variables.md?plain=1#L115-L117)

So, I think for v1 compat, this makes sense! But we'll definitely need to take care to find any remaining errant references to the order being the other way around, and given that v2 is now flip-flopping its behavior, should make sure we call that out explicitly in release notes, along with the "workaround" (use `MYVAR=${MYVAR:-DefaultValue}` syntax in `.env` file), and explanation that it should now behave the same between v1+v2 😅

**Review by milas — APPROVED** [review]

**Comment by milas** [comment]
@ulyssessouza @glours I found a case where (after these changes) v1 and v2 behave slightly differently: the _in-file_ interpolation in v1 _DID_ take shell as higher precedence than env.

Arguably, that's a bug/inconsistency in v1, but noting here so that we can mention it in docs/release notes somehow and to make sure there's not some reason that behavior _was_ desirable???

```shell
$ cat .env
VAR=dotenv
```

```shell
$ cat docker-compose.yaml
services:
  echo:
    image: alpine
    command: [
      '/bin/ash',
      '-c',
      'echo "Compose config: ${VAR} | Container: $${VAR}"'
    ]
    env_file: .env
```

| Version         | Command                       | Result                                                             |
|-----------------|-------------------------------|--------------------------------------------------------------------|
| v1              | `docker-compose up`           | Compose config: dotenv &#124; Container: dotenv                    |
| v1              | `VAR=shell docker-compose up` | Compose config: shell    &#124;                  Container: dotenv |
| v2 (with fixes) | `docker compose up`           | Compose config: dotenv     &#124; Container: dotenv                |
| v2 (with fixes) | `VAR=shell docker compose up` | Compose config: dotenv       &#124; Container: dotenv              |

**Comment by ulyssessouza** [comment]
@milas Hmmm! It's because in V1 the precedence of `.env` and the OS Env are switched in this case. So the OS is not overwritten by the `.env`.

**Comment by milas** [comment]
> in V1 the precedence of `.env` and the OS Env are switched in this case

Yeah, I'd argue this is a bug / inconsistency / undesirable behavior from v1, so I don't think it makes sense to carry over.

I think the behavior of v2 (with your changes from this PR + others) is now good/correct! 💪

**Comment by glours** [comment]
🎉

**Pull request merged**
